### PR TITLE
Avoid zombie invasion [v2]

### PR DESCRIPTION
**Pull request by ldoktor**
Actually I already found out the https://github.com/avocado-framework/avocado/pull/2945 implementation is broken and can lead to zombie invasion...

v1: https://github.com/avocado-framework/avocado/pull/2954

Changes:

```yaml
v2: Add selftest for return value and check for it in existing tests
```

### CODE REVIEW

File: avocado/utils/process.py [hunk 1]
```diff
@@ -209,36 +209,41 @@ def kill_process_tree(pid, sig=signal.SIGKILL, send_sigcont=True,
 
     :param pid: The pid of the process to signal.
     :param sig: The signal to send to the processes.
+    :param send_sigcont: Send SIGCONT to allow killing stopped processes
     :param timeout: How long to wait for the pid(s) to die
                     (negative=infinity, 0=don't wait,
                     positive=number_of_seconds)
+    :return: killed_pids that were destroyed
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply avocado/utils/process.py]: Oh, and while at it, it'd be nice to make it extra clear that the return value is always a list (may be useful given that the `pid` parameter is not a list).

A simple `:rtype: list` does the trick.

File: avocado/utils/process.py [hunk 2]
```diff
@@ -209,36 +209,41 @@ def kill_process_tree(pid, sig=signal.SIGKILL, send_sigcont=True,
...
-        start = time.time()
+    def _all_pids_dead(killed_pids):
+        for pid in killed_pids:
+            if pid_exists(pid):
+                return False
+        return True
 
-        def remaining_time():
-            return timeout - time.time() + start
+    killed_pids = [pid]
```
Comment (clebergnu) [inline avocado/utils/process.py]: I understand the optimization here, but the name of this variable is misleading: up to this point, we cannot tell if any process has been killed -- clearly none was killed on our behalf.

Something like this:

```diff
diff --git a/avocado/utils/process.py b/avocado/utils/process.py
index 5296ef4a..f5029d8e 100644
--- a/avocado/utils/process.py
+++ b/avocado/utils/process.py
@@ -221,12 +221,12 @@ def kill_process_tree(pid, sig=signal.SIGKILL, send_sigcont=True,
                 return False
         return True
 
-    killed_pids = [pid]
     if timeout > 0:
         start = time.time()
 
     if not safe_kill(pid, signal.SIGSTOP):
-        return killed_pids
+        return [pid]
+    killed_pids = [pid]
     for child in get_children_pids(pid):
         killed_pids.extend(kill_process_tree(int(child), sig, False))
     safe_kill(pid, sig)
```

Makes reading the code straightforward IMO.

Comment (ldoktor) [reply avocado/utils/process.py]: ok

File: avocado/utils/process.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ldoktor) [reply avocado/utils/process.py]: sure

**Comment by ldoktor** [comment]
[Allow waiting for all pids to die in "process.kill_process_tree".](https://trello.com/c/5e37RsxH/1468-allow-waiting-for-all-pids-to-die-in-processkillprocesstree)

**Pull request closed without merging**